Provide more extensive file filtering features

weierophinney opened the pull request:
|    Q        |   A
|------------ | ------
| Bugfix      | no
| BC Break    | no
| New Feature | yes
| RFC         | no

### Description

This patch provides a number of improvements as suggested by several users.

First, it adds full documentation of the command when invoked using `laminas-migration help migrate`, complete with examples.

Second, it updates the `--exclude` option to also allow providing filenames to exclude.

Third, it adds new options:

- `--no-config-processor` will skip injection of either the laminas-zendframework-bridge module or config post processor if the tool detects that an injection would otherwise be made.

- `--filter|-f` can be invoked multiple times, and allows providing a regular expression (minus any delimiters). If a file does not match any filters provided, it will not be processed. (If it matches at least one filter, and no exclude rules match, it will be processed.)

weierophinney committed:
feat: more docs and more options

- Adds full command documentation.
- Adds the options:
  - --exclude-file|-x (exclude an individual file; allow multiple invocations)
  - --filter|-f (provide a regex filter to satisfy; allow multiple invocations)
  - --no-config-processor (skip injection of the Module or ConfigPostProcessor)

weierophinney committed:
feat: implement functionality to skip injection of module, config post processor

Adds functionality to `injectBridgeModule` and
`injectBridgeConfigPostProcessor` to return early if the file was
detected, but the `--no-config-processor` switch was provided.

weierophinney committed:
feat: add functionality to provide regexp file filters

Any file that does not match the filter will be excluded from rewriting.

weierophinney committed:
feat: add <info> tags around headings in help text

michalbundyra commented on src/Command/MigrateCommand.php:
```diff
@@ -22,10 +22,80 @@
+or "--exclude data/cache". The --exclude option can be issued multiple times,
+one for each directory you wish to exclude.
+
+To exclude individual files, use the --exclude-file (-x) option. This option
```
I'd prefer to have just `--exclude` which will accept either directories and files, so we can do:
```--exclude data/cache --exclude dump.sql```

it will simplify user interface.

weierophinney replied:
I agree, but there are some logistical issues I need to solve before I can say one way or the other if it's possible. I'll attempt those shortly.

weierophinney replied:
Figured out a solution!

michalbundyra commented on src/Command/MigrateCommand.php:
```diff
@@ -231,10 +318,66 @@ static function (SplFileInfo $current, $key, $iterator) use ($exclude) {
+         */
+        $exclusions[] = static function ($path) use ($filters) {
+            foreach ($filters as $filter) {
+                $pattern = sprintf('#%s#', $filter);
```
I believe we should use `preg_quote` (with delimiter, 2nd arg)

weierophinney replied:
I initially did, but this led to all sorts of issues. Dots, slashes, parens, pipes, brackets, and braces were all quoted, which meant that any attempts at quantifiying, using character classes, or grouping would fail.

weierophinney replied:
Figured out a solution!

michalbundyra commented on src/Command/MigrateCommand.php:
```diff
@@ -231,10 +318,66 @@ static function (SplFileInfo $current, $key, $iterator) use ($exclude) {
+     * @param string[] $excludePaths
+     * @param string $path
+     * @param callable[] $exclusions
+     * @return array The $exclusions array with the new exclusion filter
```
`callable[]`  instead of `array`?

michalbundyra commented on src/Command/MigrateCommand.php:
```diff
@@ -255,14 +398,46 @@ private function createExclusionChecker($path, array $excludePaths)
+     * @param string[] $excludePaths
+     * @param string $path
+     * @param callable[] $exclusions
+     * @return array The $exclusions array with the new exclusion filter
```
`callable[]` instead of `array`

weierophinney replied:
No longer relevant with the most recent refactoring.

michalbundyra commented on src/Command/MigrateCommand.php:
```diff
@@ -282,15 +457,21 @@ private function createNewDirectory($path)
+     * @param bool $disableConfigProcessorInjection
      */
-    private function injectBridgeModule($path, SymfonyStyle $io)
+    private function injectBridgeModule($path, $disableConfigProcessorInjection, SymfonyStyle $io)
```
yday I've added support for Symfony Console 2.1+ but ... we need to check if it works with SymfonyStyle... I haven't checked it, tbh

weierophinney replied:
They were added in 2.7. As such, that needs to be our minimum version.

I think that's fairly reasonable, as it's 4.5 years old.

weierophinney committed:
refactor: rename createExclusionChecker to createFilter

Better details what the purpose is.

weierophinney committed:
refactor: have filter follow semantics of array_filter

The way array_filter works is that if the callback returns `false`, the
item is dropped from the array. As such, I've reversed how the callbacks
created by each of createFileFilter, createDirectoryExclusionChecker,
and createFileExclusionChecker work, with a corresponding change in
findProjectFiles.

weierophinney committed:
refactor: merge --exclude and --exclude-file

This patch merges the `--exclude` and `--exclude-file` flags, allowing
usage of `--exclude` to work for both directories and files.

Additionally, I've refactored the internal logic for how the various
filters are aggregated, and how the return values are used.

weierophinney committed:
feat: require symfony/console 2.7 as minimum version

First version with SymfonyStyle class

weierophinney committed:
fix: static closures

weierophinney committed:
fix: use DIRECTORY_SEPARATOR when creating directory exclusion patterns

For cross-platform compatibility.

weierophinney committed:
fix: use correct PCRE delimiter when quoting

michalbundyra reviewed (approved):
🚢

weierophinney committed:
qa: adds documentation about why preg_quote is and isn't used.
